server/app/services/grid_service_instance_deployer.rb

```
@@ -27,68 +30,51 @@ def deploy(node, instance_number, deploy_rev, creds = nil)
+  def stop_current_instance(current_instance)
+    current_instance.set(desired_state: 'stopped')
+    if current_instance.connected?
+      notify_node(current_instance.host_node)
+      wait_for_service_state(current_instance, 'stopped')
     end
   end
 
-  # @param [HostNode] node
-  # @param [String] instance_number
-  # @return [Boolean]
-  def service_exists_on_node?(node, instance_number)
-    old_container = self.grid_service.containers.service_instance(
-      self.grid_service, instance_number
-    ).first
-    old_container && old_container.host_node && old_container.host_node == node
+  # @param [GridServiceInstance] service_instance
+  def wait_for_service_state(service_instance, state)
+    Timeout.timeout(300) do
+      until service_instance.reload.state == state do
+        sleep 1.0
+      end
+    end
   end
 
   # @param [HostNode] node
   # @param [String] instance_number
   # @param [String] deploy_rev
-  # @param [Hash, NilClass] creds
-  def create_service_instance(node, instance_number, deploy_rev, creds)
-    creator = Docker::ServiceCreator.new(self.grid_service, node)
-    creator.create_service_instance(instance_number, deploy_rev, creds)
-  end
-
-  # @param [String] instance_number
-  # @param [String] deploy_rev
-  # @return [Boolean]
-  def deployed_service_container_exists?(instance_number, deploy_rev)
-    container = self.find_service_instance_container(instance_number, deploy_rev)
-    if container && container.container_id
-      true
-    else
-      false
+  # @return [GridServiceInstance]
+  def create_service_instance(node, instance_number, deploy_rev)
+    instance = current_service_instance(instance_number)
+    unless instance
+      instance = GridServiceInstance.create!(
+        host_node: node, grid_service: self.grid_service, instance_number: instance_number
+      )
     end
+    instance.set(host_node_id: node.id, deploy_rev: deploy_rev, desired_state: 'running')
+
+    instance
+  rescue => exc
```

This will end badly, just let this propagate up to `deploy`?